be/src/runtime/runtime_filter_cache.cpp

```diff
@@ -0,0 +1,217 @@
+// This file is licensed under the Elastic License 2.0. Copyright 2021-present, StarRocks Limited.
+
+#include "runtime/runtime_filter_cache.h"
+
+#include <chrono>
+
+#include "gutil/strings/substitute.h"
+#include "util/thread.h"
+
+namespace starrocks {
+
+using std::chrono::seconds;
+using std::chrono::milliseconds;
+using std::chrono::steady_clock;
+using std::chrono::duration_cast;
+class RfCacheValue {
+public:
+    RfCacheValue() { _extend_lifetime(); }
+    ~RfCacheValue() = default;
+    void put_if_absent(int filter_id, const JoinRuntimeFilterPtr& filter) {
+        _extend_lifetime();
+        if (!_filters.count(filter_id)) {
+            _filters[filter_id] = filter;
+        }
+    }
+
+    JoinRuntimeFilterPtr get(int filter_id) const {
+        _extend_lifetime();
+        auto it = _filters.find(filter_id);
+        if (it != _filters.end()) {
+            return it->second;
+        } else {
+            return nullptr;
+        }
+    }
+
+    bool is_expired() const {
+        auto now = duration_cast<milliseconds>(steady_clock::now().time_since_epoch()).count();
+        return now > _deadline;
+    }
+
+private:
+    void _extend_lifetime() const {
+        _deadline = duration_cast<milliseconds>(steady_clock::now().time_since_epoch() + EXPIRE_SECONDS).count();
+    }
+    std::unordered_map<int, std::shared_ptr<vectorized::JoinRuntimeFilter>> _filters;
+    mutable int64_t _deadline;
+    static constexpr seconds EXPIRE_SECONDS = seconds(60);
+};
+
+class RfEvent {
+public:
+    RfEvent(const TUniqueId& query_id, int filter_id, std::string&& msg)
+            : query_id(query_id),
+              filter_id(filter_id),
+              ts(duration_cast<milliseconds>(steady_clock::now().time_since_epoch()).count()),
+              msg(std::move(msg)) {}
+    ~RfEvent() = default;
+    std::string to_string() { return strings::Substitute("[$0][$1] $2: $3", print_id(query_id), filter_id, ts, msg); }
+
+private:
+    TUniqueId query_id;
+    int filter_id;
+    long ts;
+    std::string msg;
+};
+using RfEventPtr = std::shared_ptr<RfEvent>;
+using RfEvents = std::list<RfEventPtr>;
+class RfEventValue {
+public:
+    RfEventValue() = default;
+    ~RfEventValue() = default;
+    void add_rf_event(const TUniqueId& query_id, int filter_id, std::string&& msg) {
+        _extend_lifetime();
+        _events.push_back(std::make_shared<RfEvent>(query_id, filter_id, std::move(msg)));
```

Comment:
If there are a lot of events, I suggest to design better data structure. `std::list + make_shared<RfEvent>`, I'm not sure if it's high performance or not.

Reply:
high performance is needless here, this feature is used for developer to inspect the performance bottleneck of grf delivery. in default, this feature is off, only when you suspect the grf delivery have some problem, you can turn on it to found out the result in trace information.

Reply:
I understand high performance is unecessary. But changing `std::list` to `std::vector` does not require much changes, but make performance better. Why not do it?